### PR TITLE
Switch list vs grid

### PR DESCRIPTION
**Pull request by tobiasKaminsky**
As discussed in #1081

I have used SharedPreferences for this and not a sqlite db.

Should we have a non blockable information on first change like this "the view settings is remembered for this and all subfolder, if they are not set otherwise".

@davivel @jancborchardt 

(minor side note: I had accidental pushed the commits direct to develop, but have reverted this)
Sorry for this...

### CODE REVIEW

**Comment by tobiasKaminsky** [comment]
Another point is: do we have to care about deleting entries in the SharedPreferences? And if so, how do we do it?

**Comment by jancborchardt** [comment]
Please review also @AndyScherzinger @Stoyicker @Kernald :)

File: src/com/owncloud/android/utils/DisplayUtils.java [hunk 1]
```diff
@@ -373,4 +376,65 @@ public static Point getScreenSize(Activity caller) {
...
+        }
+    }
+
+    public static void setViewMode(OCFile file, boolean setGrid){
+        SharedPreferences setting = MainApp.getAppContext().getSharedPreferences(
+                "viewMode", Context.MODE_PRIVATE);
+
+        SharedPreferences.Editor editor = setting.edit();
+        editor.putBoolean(file.getRemoteId(), setGrid);
+        editor.commit();
```
Comment (Kernald) [inline src/com/owncloud/android/utils/DisplayUtils.java]: Synchronicity is not important here, you should call `editor.apply()` instead of `editor.commit()`

**Comment by Kernald** [comment]
Removing an entry in `SharedPreferences` is really simple: http://developer.android.com/reference/android/content/SharedPreferences.Editor.html#remove(java.lang.String)

We should probably remove corresponding entries when a folder is removed. The back-end file is loaded once when it's first needed then kept in memory, if the user switch the view mode in a lot of folders, the initial load could take long. I never experienced this kind of issue, and don't really know if the impact is really important.

Another idea to keep this file the smallest possible: instead of storing both list and grid state, remove the entry when the user switches back to the default (list).

**Comment by tobiasKaminsky** [comment]
With this implemention there is only a switch between grid and list view possible.
But it is not possible to manually delete the setting for a specific folder.
So maybe it should be something like that:
- grid
- list
- inherited
And the third option would delete the specific setting.

**Comment by AndyScherzinger** [comment]
Looks good to me, only one thing: I would recommed/prefer to switch the menu icon so you wouldn't need the text (done like this in Google's apps as well imho) to distinguish the action but could (also) show it visually :smiley: 

so I would recommend using:
For show list view
https://www.google.com/design/icons/#ic_view_list
and for grid
https://www.google.com/design/icons/#ic_view_module

**Comment by tobiasKaminsky** [comment]
Do you mean, that the action is always visible? (Like currently "upload file" and "create folder")?
For me there is no icon shown in the menu next to the text.

But these are great icons.

switch

**Comment by AndyScherzinger** [comment]
Yes, I meant always shown but that probably only makes sense after my fab speed dial implementation has been tested and merged and thus the space in the action bar has been freed up. 
Am 19.08.2015 19:28 schrieb tobiasKaminsky <notifications@github.com>:Do you mean, that the action is always visible? (Like currently "upload file" and "create folder")?
For me there is no icon shown in the menu next to the text.

But these are great icons.

**Comment by tobiasKaminsky** [comment]
@AndyScherzinger 
I have put those files in the corresponding folders: 
drawable
drawable-mdpi
drawable-hdpi
drawable-xdpi
But still the xml code says: no file:
 < item
        android:id="@+id/action_switch_view"
        android:icon="@android:drawable/ic_view_list"
        android:orderInCategory="2"
        app:showAsAction="always"
        android:title="@string/action_switch_grid_view" />

Do you have any idea?

**Comment by Kernald** [comment]
@tobiasKaminsky : it should be `@drawable/ic_view_list` instead of `@android:drawable/ic_view_list` if the file is in the application's folders.

By the way, the `drawable` folder without density qualification should not contain PNG assets, only scale-independent ones (XML variants of PNG assets, shapes...).

**Comment by Kernald** [comment]
And as a side note: in addition to xhdpi, we probably should provide xxhdpi assets, most of high end devices are in this density bucket now.

**Comment by tobiasKaminsky** [comment]
drawer2
drawer1

@Kernald Thanks for your hint.
One more question: https://www.google.com/design/icons/#ic_view_list has the icon in 18 24 36 48dp.
How do I know which size has to be in which drawable-??? folder?

**Comment by Kernald** [comment]
@tobiasKaminsky : if you download, for example the 24dp version, you'll see that you get a zip with the corresponding drawable-bucket folders. The question is: where do you plan to put your icon? For an `ActionBar`/`Toolbar`, as mentioned [here](https://www.google.com/design/spec/layout/metrics-keylines.html#metrics-keylines-touch-target-size), you should use the 24dp one.

**Comment by jancborchardt** [comment]
@tobiasKaminsky I’m not sure we need to put the switcher that present.

**Comment by tobiasKaminsky** [comment]
You are the design/UI guy  ;)
So it is up to you.

**Comment by jancborchardt** [comment]
@tobiasKaminsky hehe, ok :) I would put it in the dot-dot-dot menu for now. :)

**Comment by tobiasKaminsky** [comment]
As you wish :) :)

**Comment by AndyScherzinger** [comment]
Sounds good to me too. I would recommend putting it on the bar out of the overflow menu as soon as  (or in case) we finish/merge/agree-on the FAB implementation since the upload/create actions would then be removed from the action bar.

File: src/com/owncloud/android/ui/activity/FileDisplayActivity.java [hunk 2]
```diff
@@ -1200,6 +1246,15 @@ public void onBrowsedDownTo(OCFile directory) {
         cleanSecondFragment();
         // Sync Folder
         startSyncFolderOperation(directory, false);
+
+        MenuItem menuItem = mOptionsMenu.findItem(R.id.action_switch_view);
+
+        changeGridIcon();
+        if (DisplayUtils.isGridView(directory, getStorageManager())){
```
Comment (AndyScherzinger) [inline src/com/owncloud/android/ui/activity/FileDisplayActivity.java]: same code fragment as line 325ff

**Comment by AndyScherzinger** [comment]
Code review done! :+1:

File: src/com/owncloud/android/ui/activity/FileDisplayActivity.java (hunk 2 above)
Comment (tobiasKaminsky) [inline src/com/owncloud/android/ui/activity/FileDisplayActivity.java]: Right!
I have moved this to a separate function.
Thanks!

**Comment by davivel** [comment]
Reviewing this now.

**Comment by davivel** [comment]
> I have used SharedPreferences for this and not a sqlite db.

Why?

**Comment by AndyScherzinger** [comment]
I guess both ways would work (preferences Vs. DB) but he probably considered this being a setting/preference and thus used SharedPreferences to implement this :)

**Comment by davivel** [comment]
> - grid
> - list
> - inherited

That third option seems created just for techies. I can imagine my mother browsing their files in the OC app and switching between grid and list without no fear, but sure she would call me by phone to wonder what does that of 'inherited' mean. And I'm not sure I would get to explain it to her.

@jancborchardt , what do you think of this?

File: src/com/owncloud/android/ui/activity/FileDisplayActivity.java [hunk 3]
```diff
@@ -321,6 +322,8 @@ private void initFragmentsWithFile() {
                     startTextPreview(file);
             }
 
+            switchLayout(getFile());
```
Comment (davivel) [inline src/com/owncloud/android/ui/activity/FileDisplayActivity.java]: There is a local variable ```file``` set to ```getFile()``` 13 lines above; I would use it instead of calling ```getFile()``` again.

**Comment by tobiasKaminsky** [comment]
The third option is not valid anymore.
There is only "switch to grid view" or "switch to list view"

**Comment by tobiasKaminsky** [comment]
SharedPreferences was a lot easier to implement as I know how to do it, whereas using a sqlite db is new to me.
If you think that it is better to use a db, I/we can try to change it.

File: src/com/owncloud/android/ui/activity/FileDisplayActivity.java (hunk 3 above)
Comment (tobiasKaminsky) [inline src/com/owncloud/android/ui/activity/FileDisplayActivity.java]: As it is only a quick change, maybe you can do it, if it otherwise will prevent this PR from QA?
(I can do it on sunday or monday afternoon)

**Comment by AndyScherzinger** [comment]
I can also sidewing an sqlite implementation in case it will be decided to go for a db based solution

File: src/com/owncloud/android/ui/activity/FileDisplayActivity.java [hunk 4]
```diff
@@ -1200,6 +1250,11 @@ public void onBrowsedDownTo(OCFile directory) {
         cleanSecondFragment();
         // Sync Folder
         startSyncFolderOperation(directory, false);
+
+        MenuItem menuItem = mOptionsMenu.findItem(R.id.action_switch_view);
```
Comment (davivel) [inline src/com/owncloud/android/ui/activity/FileDisplayActivity.java]: Variable not used; please, remove this line.

File: src/com/owncloud/android/ui/activity/FileDisplayActivity.java [hunk 5]
```diff
@@ -497,6 +509,12 @@ public boolean onPrepareOptionsMenu(Menu menu) {
     public boolean onCreateOptionsMenu(Menu menu) {
         MenuInflater inflater = getMenuInflater();
         inflater.inflate(R.menu.main_menu, menu);
+        mOptionsMenu = menu;
+
+        MenuItem menuItem = mOptionsMenu.findItem(R.id.action_switch_view);
```
Comment (davivel) [inline src/com/owncloud/android/ui/activity/FileDisplayActivity.java]: Variable not used; please, remove this line.

File: src/com/owncloud/android/ui/fragment/ExtendedListFragment.java [hunk 6]
```diff
@@ -127,6 +127,13 @@ protected void switchToListView() {
             mCurrentListView = mListView;
         }
     }
+
+    public boolean isGridView(){
```
Comment (davivel) [inline src/com/owncloud/android/ui/fragment/ExtendedListFragment.java]: Very minor. Why the change from ```isGridMode()``` to ```isGridView()```?

File: src/com/owncloud/android/ui/fragment/FileDetailFragment.java [hunk 7]
```diff
@@ -217,6 +217,24 @@ public void onPrepareOptionsMenu(Menu menu) {
             item.setVisible(false);
             item.setEnabled(false);
         }
+
+        item = menu.findItem(R.id.action_switch_view);
```
Comment (davivel) [inline src/com/owncloud/android/ui/fragment/FileDetailFragment.java]: This shouldn't be this way. In tablets in landscape mode, when a file is shown in the right pane, the options for the current folder still make sense. Please, undo this.

File: src/com/owncloud/android/ui/preview/PreviewTextFragment.java [hunk 8]
```diff
@@ -290,6 +290,18 @@ public void onPrepareOptionsMenu(Menu menu) {
             item.setVisible(false);
             item.setEnabled(false);
         }
+
+        item = menu.findItem(R.id.action_switch_view);
```
Comment (davivel) [inline src/com/owncloud/android/ui/preview/PreviewTextFragment.java]: Same comment here as in ```FileDetailFragment```

File: src/com/owncloud/android/datamodel/OCFile.java [hunk 9]
```diff
@@ -74,6 +75,8 @@ public OCFile createFromParcel(Parcel source) {
 
     private boolean mIsDownloading;
 
+    private boolean mShowGridView;
```
Comment (davivel) [inline src/com/owncloud/android/datamodel/OCFile.java]: Never used, right? Please, remove it.

File: src/com/owncloud/android/utils/DisplayUtils.java [hunk 10]
```diff
@@ -245,6 +258,67 @@ public static Point getScreenSize(Activity caller) {
...
+            return !setting.contains(parentDir.getRemoteId());
+        }
+    }
+
+    public static void setViewMode(OCFile file, boolean setGrid){
+        SharedPreferences setting = MainApp.getAppContext().getSharedPreferences(
+                "viewMode", Context.MODE_PRIVATE);
+
+        SharedPreferences.Editor editor = setting.edit();
+        editor.putBoolean(file.getRemoteId(), setGrid);
```
Comment (davivel) [inline src/com/owncloud/android/utils/DisplayUtils.java]: I'd rather use ```file.getId()``` as the index. In theory ```file.getRemoteId()``` has information enough to not collide even through different server, but ```file.getId()``` does not depend on theory, it's unique for sure.

**Comment by davivel** [comment]
> The third option is not valid anymore.

I'm sorry, I saw it later in the code.

> If you think that it is better to use a db, I/we can try to change it.

> I can also sidewing an sqlite implementation in case it will be decided to go for a db based solution

Thank you both. Maybe it isn't worth it. My main concern if how nice the solution will scale if a user sets different values for lots and lots of folders, but probably this is not a realistic case. There are changes more important to do in the persistence layer, so let's keep it this way. If tests reveal performance problems, will look into it again.

File: src/com/owncloud/android/utils/DisplayUtils.java [hunk 11]
```diff
@@ -245,6 +258,67 @@ public static Point getScreenSize(Activity caller) {
     }
 
     /**
+     * Determines if user set folder to grid or list view. If folder is not set itself,
+     * it finds a parent that is set (at least root is set).
+     * @param file
+     * @param storageManager
+     * @return
+     */
+    public static boolean isGridView(OCFile file, FileDataStorageManager storageManager){
```
Comment (davivel) [inline src/com/owncloud/android/utils/DisplayUtils.java]: I wouldn't do this in a static method. An ```Activity``` or a ```Fragment``` would be a better place.

Even further, did you try to implement the code added to ```FileDisplayActivity``` in ```OCFileListFragment``` instead? That would result in a better structure IMHO. The feature is contained in the scope of ```OCFileListFragment``` responsibility. 

The addition of the menu item in the ```ActionBar``` should be possible also from ```OCFileListFragment```.

Would you give it a try?

Comment (davivel) [inline src/com/owncloud/android/utils/DisplayUtils.java]: Very minor: ```isGridView``` could be confused with the meaning of ```isGridView``` (or even ```isGridMode```) in ```ExtendedListFragment```. 

What about ```isGridPrefered```, or something similar?

**Comment by davivel** [comment]
Review finished. @tobiasKaminsky , have a look to my comments, please.

Thanks.

File: src/com/owncloud/android/ui/activity/FileDisplayActivity.java (hunk 3 above)
Comment (tobiasKaminsky) [inline src/com/owncloud/android/ui/activity/FileDisplayActivity.java]: Done

File: src/com/owncloud/android/ui/activity/FileDisplayActivity.java (hunk 4 above)
Comment (tobiasKaminsky) [inline src/com/owncloud/android/ui/activity/FileDisplayActivity.java]: Done

File: src/com/owncloud/android/ui/activity/FileDisplayActivity.java (hunk 5 above)
Comment (tobiasKaminsky) [inline src/com/owncloud/android/ui/activity/FileDisplayActivity.java]: Done for the menuItem.
As I recall correctly it is important to have the mOptionMenu updated.

File: src/com/owncloud/android/ui/fragment/ExtendedListFragment.java (hunk 6 above)
Comment (tobiasKaminsky) [inline src/com/owncloud/android/ui/fragment/ExtendedListFragment.java]: Do not know anymore. Maybe just because it was implemented on different days ;)

Comment (tobiasKaminsky) [inline src/com/owncloud/android/ui/fragment/ExtendedListFragment.java]: I will change it to gridView as it is a view/layout, ok?

File: src/com/owncloud/android/ui/fragment/FileDetailFragment.java (hunk 7 above)
Comment (tobiasKaminsky) [inline src/com/owncloud/android/ui/fragment/FileDetailFragment.java]: Done

File: src/com/owncloud/android/ui/preview/PreviewTextFragment.java (hunk 8 above)
Comment (tobiasKaminsky) [inline src/com/owncloud/android/ui/preview/PreviewTextFragment.java]: Done

File: src/com/owncloud/android/datamodel/OCFile.java (hunk 9 above)
Comment (tobiasKaminsky) [inline src/com/owncloud/android/datamodel/OCFile.java]: Done

File: src/com/owncloud/android/utils/DisplayUtils.java (hunk 10 above)
Comment (tobiasKaminsky) [inline src/com/owncloud/android/utils/DisplayUtils.java]: The "problem" is that .putBoolean expects a string, so I have to convert it.
But...done ;)

**Comment by tobiasKaminsky** [comment]
I have also updated it to the latest master.

**Comment by tobiasKaminsky** [comment]
@davivel 
Everything is now included and up to date. :)

File: src/com/owncloud/android/ui/fragment/ExtendedListFragment.java (hunk 6 above)
Comment (davivel) [inline src/com/owncloud/android/ui/fragment/ExtendedListFragment.java]: :+1:

File: src/com/owncloud/android/ui/activity/FileDisplayActivity.java (hunk 5 above)
Comment (davivel) [inline src/com/owncloud/android/ui/activity/FileDisplayActivity.java]: Yes, handling of mOptionsMenu is right.

File: src/com/owncloud/android/utils/DisplayUtils.java (hunk 10 above)
Comment (davivel) [inline src/com/owncloud/android/utils/DisplayUtils.java]: Good catch :)

File: src/com/owncloud/android/ui/activity/FileDisplayActivity.java [hunk 12]
```diff
@@ -567,12 +568,38 @@ public void onClick(DialogInterface dialog, int which) {
                 builder.create().show();
                 break;
             }
+            case R.id.action_switch_view:{
+                if (isGridView()){
+                    item.setTitle(getApplicationContext().getString(R.string.action_switch_grid_view));
```
Comment (davivel) [inline src/com/owncloud/android/ui/activity/FileDisplayActivity.java]: ```getAppicationContext()``` is not really needed. ```FileDisplayActivity``` is a ```Context```, so ```getString``` is available in ```this```.

Or am I missing something?

File: src/com/owncloud/android/ui/activity/FileDisplayActivity.java [hunk 13]
```diff
@@ -567,12 +568,38 @@ public void onClick(DialogInterface dialog, int which) {
...
+                return true;
+            }
         default:
             retval = super.onOptionsItemSelected(item);
         }
         return retval;
     }
 
+    public void setViewMode(OCFile file, boolean setGrid){
+        SharedPreferences setting = MainApp.getAppContext().getSharedPreferences(
```
Comment (davivel) [inline src/com/owncloud/android/ui/activity/FileDisplayActivity.java]: Same here, ```getSharedPreferences``` should be available directly in the ```FileDisplayActivity```

**Comment by davivel** [comment]
Thanks a lot, @tobiasKaminsky .

Just a couple of extra comments. Should be the last, from my side. Almost there.

**Comment by davivel** [comment]
@tobiasKaminsky , if you don't mind, I'll make those final changes by myself; this is now between the top priorities and QA is willing to start with it :) .

**Comment by davivel** [comment]
Scheduled for next release

**Comment by tobiasKaminsky** [comment]
Please feel free to do it.
I can do it in ~5 hours.

Great to see this progress :)

**Comment by davivel** [comment]
This is :+1: for me now.

@tobiasKaminsky , review my last changes.

@jesmrec , if @tobiasKaminsky is OK with the last changes, this is ready to start QA.

**Comment by tobiasKaminsky** [comment]
Just a quick look, but it looks good to me :+1:
So, QA can start :)

**Comment by davivel** [comment]
Rebasing on master...

**Comment by davivel** [comment]
This was trickier to rebase than expected; @jesmrec, part of the tests should be repeated to check that nothing was wrong

**Comment by jesmrec** [comment]
The tests were passed again succesfully.

**Comment by rperezb** [comment]
@carlaschroder new feature for the Android app, being possible to decide the list mode between grid and list one
